### PR TITLE
ICSD-1100 Send metrics to alert Rogue Consul server.

### PR DESCRIPTION
**Pull request by jzou-rbx**
Write metrics to indicate when a Consul server is identified as a rogue, which will be later used for AlertManager.
https://grafana.rbx.com/d/d026dbe0-e4d5-4d08-9bbe-4a0c922527f7/rogue-consul-server?orgId=1&var-datasource=All&var-site=snc1&from=now-12h&to=now

Follow up to https://github.com/Roblox/consul-cookbook/pull/44 (based on top of it)

### CODE REVIEW

File: recipes/server_check.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    content <<-EOF
+        [Unit]
+        ConditionPathExists=!#{drop_in_file}
+        EOF
+    action :create
+  end
+
+  # Send a binary value as metric to indicate the server did not start.
+  unless node['consul']['metrics_write_endpoint'].nil?
+    http_request 'send metrics' do
```
Comment (sherrman) [inline recipes/server_check.rb]: I wonder if writing something to the grafana-agent's textfiles is better (or something to do in addition).

This POST can fail (after 5 retries).

But textfiles are forever! I'm also unclear what this metrics call does.

grafana-agent will constantly report there's a misconfigured node which works much better with how prom/vm alerting works.

Comment (jzou-rbx) [reply recipes/server_check.rb]: Thanks Sherman, let me look up grafana-agent and how to use that.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # Change Log
 
+## 9003.1.34
```
Comment (sherrman) [inline CHANGELOG.md]: yass

**Review by sherrman — APPROVED** [review]
Awesome! 🚀

File: recipes/server_check_helper.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  end
+end
+
+def write_metric(metric_file, metric_name, metric_help, metric_value)
+  file metric_file do
+    owner 'root'
+    mode '0644'
+    content <<-EOF.gsub(/^ */, '')
+      # HELP #{metric_name} #{metric_help}
+      #{metric_name} #{metric_value}
```
Comment (fcuny-rbx) [inline recipes/server_check_helper.rb:38]: the main concern here is that if grafana agent is not running, we will not know about this. I looked at https://roblox.atlassian.net/wiki/spaces/IC/pages/2030306932/Service+Discovery+Runbook+Decommission+A+Consul+Server and I don't see a step where we disable grafana agent, and I believe the agent is running/deployed by default in the base policy (we should confirm). Do we have a scenario in mind where we would end up with a non-working grafana agent on a rogue node ? If so, what should we consider here ?

Comment (fcuny-rbx) [reply recipes/server_check_helper.rb:38]: Oh, I see that earlier you had a POST and @sherrman suggested textfile (which was also what I originally had in mind). I'm not sure I've a great suggestion, text file are easy, POST increases the complexity (you need to handle retries, settings labels, etc).

**Review by fcuny-rbx — APPROVED** [review]

File: recipes/server_check.rb [hunk 4]
```diff
@@ -4,6 +4,10 @@
 
 consul_service_folder = '/etc/systemd/system/consul.service.d'
 drop_in_file = "#{consul_service_folder}/server_start_prevention.conf" # prevent systemd from starting
+grafana_agent_folder = '/etc/grafana-agent/textfiles' # text based metric for grafana agent to push
+metric_file = "#{grafana_agent_folder}/consul_server.prom"
+metric_name = 'consul_server_allowed_to_start' # grafana agent will add the labels such as datacenteripaddress
```
Comment (glynnforrest) [inline recipes/server_check.rb:9]: typo?
```suggestion
metric_name = 'consul_server_allowed_to_start' # grafana agent will add the labels such as datacenter, ipaddress, etc
```

File: recipes/server_check_helper.rb [hunk 5]
```diff
@@ -28,3 +28,14 @@ def reload_systemd()
     code 'systemctl daemon-reload'
   end
 end
+
+def write_metric(metric_file, metric_name, metric_help, metric_value)
+  file metric_file do
+    owner 'root'
+    mode '0644'
+    content <<-EOF.gsub(/^ */, '')
+      # HELP #{metric_name} #{metric_help}
```
Comment (glynnforrest) [inline recipes/server_check_helper.rb:37]: Please add

```
# TYPE #{metric_name} gauge
```

File: recipes/server_check.rb [hunk 6]
```diff
@@ -4,6 +4,10 @@
...
+metric_help = 'A boolean variable indicating whether this consul server should be allowed to start.'
```
Comment (glynnforrest) [inline recipes/server_check.rb:10]: Super nit - prometheus doesn't have booleans. This is a gauge type.

https://prometheus.io/docs/concepts/metric_types/#gauge

**Review by glynnforrest — APPROVED** [review]
This is fantastic, thanks @jzou-rbx. Spotted some small things but I won't block approval. We tested this together and confirmed it works.

**Pull request merged**